### PR TITLE
Update ActiveRecord 'dirty' logic

### PR DESCRIPTION
**Pull request by nbey**
- Only store original value once
- Unset original value on reverted fields
- Remove record dirtyness if the only modified field is
reverted back to its original value
- When updating relationships, only set record dirty when
modifying `one-many` & `context-children` relationships.
Other relationship types will modify the record
and internally call `_setFieldValue()` which will
determine whether or not the record is actually dirty

### CODE REVIEW

**Comment by nbey** [comment]
On second thought, we could do without setting a record dirty when updating relationships, since a deep call to save will save a records' relationships, even if the record itself isn't dirty.

File: php-classes/ActiveRecord.class.php [hunk 1]
```diff
@@ -2177,10 +2177,20 @@ protected function _setFieldValue($field, $value)
 
         if ($forceDirty || !array_key_exists($columnName, $this->_record) || $this->_record[$columnName] !== $value) {
             if (array_key_exists($columnName, $this->_record)) {
-                $this->_originalValues[$field] = $originalValue;
+                // only store original value if it hasn't been previously set
+                if (!array_key_exists($field, $this->_originalValues)) {
+                    $this->_originalValues[$field] = $originalValue;
+                } elseif ($this->_originalValues[$field] == $value) { // the value was reverted back to it's original value
```
Comment (themightychris) [inline php-classes/ActiveRecord.class.php]: @nbey let's do a triple-equal here

File: php-classes/ActiveRecord.class.php [hunk 2]
```diff
@@ -2177,10 +2177,20 @@ protected function _setFieldValue($field, $value)
...
+                    unset($this->_originalValues[$field]);
+                }
             }
             $this->_record[$columnName] = $value;
-            $this->_isDirty = true;
+            // set record dirty if there are still dirty fields remaining
```
Comment (themightychris) [inline php-classes/ActiveRecord.class.php]: @nbey just set _isDirty to the expression

File: php-classes/ActiveRecord.class.php (hunk 1 above)
Comment (nbey) [inline php-classes/ActiveRecord.class.php]: Ok.

File: php-classes/ActiveRecord.class.php (hunk 2 above)
Comment (nbey) [inline php-classes/ActiveRecord.class.php]: @themightychris I thought of that, but then I saw that there was a `$forceDirty` variable that was sometimes used. If we move the logic to the `_isDirty` expression, it'll prevent one from overriding/forcing a record's dirty-ness

Comment (themightychris) [inline php-classes/ActiveRecord.class.php]:
```
if (expression) {
    thing = true;
} else {
    thing = false;
}
```
is an antipattern, it's exactly functionally equivalent to
```
thing = expression;
```
it couldn't produce any different result as the current code if you just move the whole `if (...)` expression to the right side of the equal

Comment (nbey) [inline php-classes/ActiveRecord.class.php]: My mistake, I misread that. I thought you said set `$Record->isDirty`, not `$Record->_isDirty` to that expression

**Pull request merged**